### DISCUSSION TITLE
Trying to understand IfcSegmentedReferenceCurve

### DISCUSSION THREAD

**Discussion by RickBrice** [IfcOpenShell/IfcOpenShell#3935] [General]
I'm trying to break down and understand the definition of an [IfcSegmentedReferenceCurve](https://standards.buildingsmart.org/IFC/RELEASE/IFC4_3/HTML/lexical/IfcSegmentedReferenceCurve.htm). It looks like there were some late nights, after hours cocktail sessions involved in scripting the prose.

> A deviating explicit position of a curve segment ([IfcCurveSegment](https://standards.buildingsmart.org/IFC/RELEASE/IFC4_3/HTML/lexical/IfcCurveSegment.htm).Placement) from the axis of the base curve produces a superelevation i.e. depression or elevation from the axis of the base curve.

I think this refers to the 1/2 cant shown in ![](https://standards.buildingsmart.org/IFC/RELEASE/IFC4_3/HTML/figures/ifcsegmentedreferencecurve.jpg). Based on the business logic, where there is a left and right cant and a rail head distance to define the elevation deviation and cross slope, it seems that 1/2 cant is a special case of only one rail being elevated/depressed (rotation about left or right rail). If one rail is lowered and the other raised (rotation about the centerline), the elevation change from the axis of the base curve would be zero (or more generally any value if there was an arbitrary point of rotation).

In general I think it means the change in IfcCurveSegment.Placement.Location along the length of the segment is defined by the type and geometry of the IfcCurveSegment.ParentCurve. It is the distance between the actual curve from the IfcGradientCurve, 

For example, if the .ParentCurve is an IfcLine as IfcLine.Pnt = IfcCartesianPoint(0,5) and IfcLine.Dir = IfcDirection(1.0,0.05) the actual curve starts 5 above the IfcGradientCurve at the start of the segment and ends 5.0 + 0.05*IfcCurveSegment.SegmentLength at the end of the segment.

None of this has to do with the superelevated cross slope. That is defined by the next sentence.

>The superelevation rate of change is directly proportionate to the curve segment parent curve curvature gradient equation ([IfcCurveSegment](https://standards.buildingsmart.org/IFC/RELEASE/IFC4_3/HTML/lexical/IfcCurveSegment.htm).ParentCurve) in the linear parameter space of the base curve.

This is definitely a post-cocktail sentence. What is the parent curve's "curvature gradient equation"? 

I think the "curvature gradient equation" is analogous to IfcAlignmentCantSegment.PredefinedType and it defines how the IfcCurveSegment.Placement.Axis rotates over the length of the segment.

If the IfcCurveSegment.ParentCurve is a IfcLine and a line has infinite curvature, how does the IfcCurveSegment.Placement.Axis rotate over the length of the segment? Should it be taken as constant (corresponding to IfcAlignmentCantSegmentTypeEnum.CONSTANTCANT.)? If so, what is the value of the constant?

The [Linear Placement of Signal](https://standards.buildingsmart.org/IFC/RELEASE/IFC4_3/HTML/annex_e/alignment-geometries-and-linear-positioning/linear-placement-of-signal.html) example referenced from the IfcSegmentedReferenceCurve documentation, one of the IfcSegmentedGradientCurve's IfcCurveSegment.ParentCurve is an IfcClothoid. The curvature of a clothoid varies linearly with distance along the curve so is the 'curvature gradient equation' = d(k)/du = sqrt(PI)/|A| where k = A*s/|A^3| and s = A*u*sqrt(PI). (Equations taken from [IfcClothoid](https://standards.buildingsmart.org/IFC/RELEASE/IFC4_3/HTML/lexical/IfcClothoid.htm)). Does this correspond to IfcAlignmentCantSegmentTypeEnum.CONSTANTCANT. or IfcAlignmentCantSegmentTypeEnum.LINEARTRANSITION.?

How is this constant used to vary IfcCurveSegment.Placement.Axis along the length of the segment? If IfcCurveSegment.Placement.Axis starts at (0,0,1), what is .Axis at u = .SegmentLength?

> If no deviation in the position of the curve segment to the base curve axis is specified, the axes (Axis and RefDirection) directions of [IfcAxis2Placement](https://standards.buildingsmart.org/IFC/RELEASE/IFC4_3/HTML/lexical/IfcAxis2Placement.htm) are interpolated between the initial curve segment placement and the placement of the subsequent curve segment.

How is "no deviation in the position of the curve segment to the base curve axis is specified" possible? IfcCurveSegment.Placement is a required parameter. Perhaps it means if IfcCurveSegment.Placement.Location is (0,0,0) so the Y-deviation is zero. But this doesn't make sense given the previous sentences in the definition of IfcSegmentedReferenceCurve. The variation of Axis is "directly proportionate to the curve segment parent curve curvature gradient equation" and IfcCurveSegment.ParentCurve is required.

The idea of linearly interpolating IfcCurveSegment.Placement.Axis along the length of the segment, between the Placement.Axis defined for a segment and defined for the next segment is easy to do. The direction (either Axis or RefDirection) is (x = x1 + u(x2-x1)/L, y = y1 + u(y2-y1)/L, z = z1 + u(z2-z1)/L). The circumstances when this is required doesn't seem possible.

Finally, there aren't any specific definitions for IfcCurveSegment.Placement.RefDirection except for the case defined above (derived from the next segment's .Placement). From the figure, I might guess that .RefDirection is tangent to the IfcCurveSegment.ParentCurve.

Sorry for the brain dump - I hope you understand my ramblings and have some insights to share so I can make sense of IfcSegmentedReferenceCurve geometry.

@aothms @civilx64

**Comment by civilx64** [comment]
@RickBrice by and large I agree with what you have laid out here, although I can neither deny nor confirm the amount of late nights and cocktails.

>  it seems that 1/2 cant is a special case of only one rail being elevated/depressed (rotation about left or right rail). 

In my (admittedly minimal) rail domain experience this is the normal case.  The low railhead stays at the profile grade and the upper rail rotates to provide the necessary cant.

> I think the "curvature gradient equation" is analogous to IfcAlignmentCantSegment.PredefinedType and it defines how the IfcCurveSegment.Placement.Axis rotates over the length of the segment.

I agree.

> If the IfcCurveSegment.ParentCurve is a IfcLine and a line has infinite curvature, how does the IfcCurveSegment.Placement.Axis rotate over the length of the segment? Should it be taken as constant (corresponding to IfcAlignmentCantSegmentTypeEnum.CONSTANTCANT.)? If so, what is the value of the constant?

It depends on the value of `IfcLine.Dir`. `IfcDirection(1.0, 0.0)` corresponds to `CONSTANTCANT` and the constant is the `y` value of `IfcLine.Pnt`. `IfcLine.Dir` of `IfcDirection(1.0, g)` corresponds to `LINEARTRANSITION` calculated as you have described:

> For example, if the .ParentCurve is an IfcLine as IfcLine.Pnt = IfcCartesianPoint(0,5) and IfcLine.Dir = IfcDirection(1.0,0.05) the actual curve starts 5 above the IfcGradientCurve at the start of the segment and ends 5.0 + 0.05*IfcCurveSegment.SegmentLength at the end of the segment.   

> The [Linear Placement of Signal](https://standards.buildingsmart.org/IFC/RELEASE/IFC4_3/HTML/annex_e/alignment-geometries-and-linear-positioning/linear-placement-of-signal.html) example referenced from the IfcSegmentedReferenceCurve documentation, one of the IfcSegmentedGradientCurve's IfcCurveSegment.ParentCurve is an IfcClothoid. 

Which entity ID?

>The curvature of a clothoid ...
> Does this correspond to IfcAlignmentCantSegmentTypeEnum.CONSTANTCANT. or IfcAlignmentCantSegmentTypeEnum.LINEARTRANSITION.?

I believe it is neither.  Before looking at the docs I had assumed that `CLOTHOID` was included in the `IfcAlignmentCantSegmentType` enumeration but it is not.  Nor is `CUBIC` which is often a "good enough" approximation of the clothoid.  Therefore my interpretation is that `IfcClothoid` as the base curve of an `IfcSegmentedReferenceCurve` segment is invalid according to the spec.
 
> How is this constant used to vary IfcCurveSegment.Placement.Axis along the length of the segment? If IfcCurveSegment.Placement.Axis starts at (0,0,1), what is .Axis at u = .SegmentLength?

I think this becomes moot as there is no **Base formula (Cant)** for `CLOTHOID` defined in [8.7.2.1.2](https://ifc43-docs.standards.buildingsmart.org/IFC/RELEASE/IFC4x3/HTML/lexical/IfcAlignmentCantSegmentTypeEnum.htm#8.7.2.1.2-Type-values).  Which IMO lends additional credence to the position of clothoid cant transition as invalid.

> How is "no deviation in the position of the curve segment to the base curve axis is specified" possible? IfcCurveSegment.Placement is a required parameter. Perhaps it means if IfcCurveSegment.Placement.Location is (0,0,0) so the Y-deviation is zero.

I agree with Y-deviation as zero and my understanding is that this is describing the case where the profile grade and axis of rotation are coincident.  In other words the `IfcSegmentedReferenceCurve` and the `IfcGradientCurve` are in the same location.

> But this doesn't make sense given the previous sentences in the definition of IfcSegmentedReferenceCurve. The variation of Axis is "directly proportionate to the curve segment parent curve curvature gradient equation" and IfcCurveSegment.ParentCurve is required.

Referring back to the diagram, with this case there is no variation in Axis because the placement is not going up or down... because it is at the axis of rotation.

> The idea of linearly interpolating IfcCurveSegment.Placement.Axis along the length of the segment, between the Placement.Axis defined for a segment and defined for the next segment is easy to do. The direction (either Axis or RefDirection) is (x = x1 + u(x2-x1)/L, y = y1 + u(y2-y1)/L, z = z1 + u(z2-z1)/L). The circumstances when this is required doesn't seem possible.

This is required when the axis of rotation is at the profile grade and you therefore don't have an explicit definition of cross-slope.  Which leads to:

> Finally, there aren't any specific definitions for IfcCurveSegment.Placement.RefDirection except for the case defined above (derived from the next segment's .Placement). 

Exactly.  If rotating around the low railhead then the cant value along with `RailheadDistance` gives you cross-slope.  But if rotating about the centerline the only way to determine the cross-slope transition is by interpolating to the next segment.

> From the figure, I might guess that .RefDirection is tangent to the IfcCurveSegment.ParentCurve.

Agree, and this is what I'm calling cross-slope.

> Sorry for the brain dump - I hope you understand my ramblings and have some insights to share so I can make sense of IfcSegmentedReferenceCurve geometry.

No worries!  I find this enjoyable as the Venn diagram of civil engineers doing "heavy" software development is pretty narrow - at least in my experience.

**Reply by RickBrice** [reply]
@civilx64 Thanks for your thoughts. I'll need to process them for a while.

> Which entity ID?

Actually, there are multiple clothoids defining the cant.

#2594 = IFCSEGMENTEDREFERENCECURVE((#2595, **#2605**, #2616, **#2627**, #2638, **#2649**, #2660, **#2671**, #2682, #6201), .U., #2543, $);

#2605 = IFCCURVESEGMENT(.CONTSAMEGRADIENTSAMECURVATURE., #2609, IFCLENGTHMEASURE(0.), IFCLENGTHMEASURE(39.9999999999926), #2610);
#2610 = IFCCLOTHOID(#2613, 39.9999999999926);

#2627 = IFCCURVESEGMENT(.CONTSAMEGRADIENTSAMECURVATURE., #2631, IFCLENGTHMEASURE(0.), IFCLENGTHMEASURE(39.9999999999925), #2632);
#2632 = IFCCLOTHOID(#2635, 39.9999999999925);

#2649 = IFCCURVESEGMENT(.CONTSAMEGRADIENTSAMECURVATURE., #2653, IFCLENGTHMEASURE(0.), IFCLENGTHMEASURE(40.0000000000118), #2654);
#2654 = IFCCLOTHOID(#2657, 40.0000000000118);

#2671 = IFCCURVESEGMENT(.CONTSAMEGRADIENTSAMECURVATURE., #2675, IFCLENGTHMEASURE(0.), IFCLENGTHMEASURE(40.0000000000117), #2676);
#2676 = IFCCLOTHOID(#2679, 40.0000000000117);

**Comment by aothms** [comment]
> It looks like there were some late nights, after hours cocktail sessions involved in scripting the prose.

😄 

Let's try to come up with some suggestions then when we have verified our understanding.

> What is the parent curve's "curvature gradient equation"?

Maybe @peterrdf can rescue us once more. But I recall from overhearing conversations that cant is based on the 1st derivative of the curve.

Sorry, I can't be of much help here.

**Comment by peterrdf** [comment]
Maybe the best would be a call where we could go through the individual questions.

Just as a starter:
- IfcSegmentedReferenceCurve is a complex beast, it has two major differences towards interpretation of its segments (compared to IfcGradientCurve) and what it applies to the final curve:
- 1. the definition of the BaseCurve of the segment is indeed not the definition itself but its first derivative (therefore any IfcLine will become a constant value (therefore called CONSTANTCANT in the business logic) and IfcClothoid will become a line (therefore called LINEARTRANSITION in the business logic))
- 2. the definition of the BaseCurve has two influences, i.e. the way the super elevation is distributed along the lenght of the segment (read extra addition to z that in some cases might appear and in some cases is always 0 as start and end super elevation is 0), secondly the resulting curve from an IfcSegmentedReferenceCurve is not only defined in 3 dimensions as a point along its lengt, but it also contains some kind of tilt / rotation at each point of the curve (similar to what PCurve's have as they inherit this from the main surface they are defined on). This 'tilt' can be used for positioning of elements or for sweeps of elements. Note that in case of higher order spiral definitions the distribution of this tilt between the start and end of a segment is non-linear (while for IfcLine and IfcClothoid it is linear).


As said more than happy to have an indepth session, either in one of the weekly IF meetings or as a separate biliteral meeting.

**Reply by RickBrice** [reply]
@peterrdf I've been attending the weekly IF meetings - they are at 6am local time so I don't get all of them. A presentation on the geometry of IfcSegmentedReferenceCurve would be great. If that doesn't work for the IF leadership, I'm open to a one-on-one meeting.

**Reply by peterrdf** [reply]
Lets see what happens tomorrow, if we don;t get to it in the IF meeting we can setup a meeting next week somewhere in the evening my time (CEST + 1) what would most probably be in the morning for you (however better than 6 am :-))

**Reply by RickBrice** [reply]
Just to close the loop on this. @peterrdf provided many insights during the IF meeting last week. I understand that the change in cant slope going along the length of a curve segment is equal to the derivative of the curvature of the curve segment parent curve. I haven't had an opportunity to implement this yet, but I think it gets me on the right track. 
@aothms

**Reply by aothms** [reply]
That's great to hear!
